### ISSUE TITLE
Julia-based definition of new layer

### ISSUE THREAD

**Issue by nowozin** [dmlc/MXNet.jl#39]
Hi,

I would like to define a new layer similar to dropout, using random number generation.
According to https://mxnet.readthedocs.org/en/latest/tutorial/new_op_howto.html it is currently possible to define layers using Python.

Is it possible to define new layers using Julia, and if so, could you provide a minimal example of a Julia-defined layer?  (For example, something as simple as scaling the input.)

Thanks,
Sebastian

**Comment by pluskid** [comment]
@nowozin Unfortunately it is currently not possible yet. Because Julia GC is not thread safe currently, so that makes things a lot complicated than in the Python case. We are still working on trying to come up with a solution yet.

**Comment by nowozin** [comment]
Thank you for the quick reply and clarification.

A related question regarding the `mx.randn` functionality: I would like to generate random noise within a neural network, ideally so that the output of a `SymbolicNode` is random each time it is evaluated.  Would I have to write a new layer for this, or is there already functionality to have such random nodes?  It seems that `mx.randn` is a low-level interface that cannot be used when setting up a symbolic computation graph.

(This would be useful for variational Bayes neural network training objectives, e.g. [(Blundell et al., 2015)](http://arxiv.org/abs/1505.05424) and also described [here](http://www.nowozin.net/sebastian/blog/stochastic-computation-graphs.html).)

**Comment by vchuravy** [comment]
@nowozin I am working on this and the non functional PR is at #16

**Comment by pluskid** [comment]
There is currently no rand number generator symbolic node. But a workaround is to treat those random numbers as input data and generate from a data provider. For example, the lstm chat-rnn example uses a customized data provider to generate all-zero matrices.

**Comment by nowozin** [comment]
Thank you so much for the quick and informative replies, much appreciated.

**Comment by Andy-P** [comment]
I would like to implement a MDN loss layer and was wondering if there was any progress on this.
Also, does the above restriction apply even if one does an all Julia version?

**Comment by Andy-P** [comment]
hmm not getting a response... guess everyone is at NIPS?

**Comment by vchuravy** [comment]
Yes with NDArray it should be possible, I just have been quite busy the last two weeks and haven't gotten any coding done. Getting this to work is on the top of my priority list for the weekend.

**Comment by Andy-P** [comment]
Just to be clear, using NDArray is already possible, or will be possible soon?

**Comment by Andy-P** [comment]
@nowozin From your answer 4 days ago, it seems it is not currently possible. Can you pls confirm this.

**Comment by Andy-P** [comment]
Any progress on using NDArray + Julia to define layers?

**Comment by vchuravy** [comment]
I made progress today, but it still has some way to go, but help is always welcome.

**Comment by Andy-P** [comment]
@vchuravy  3 questions

1. Is this the branch your using to work on this?
https://github.com/vchuravy/MXNet.jl/tree/vc/nativeops/src

2. How can I help push this along?

3. Is there anything external holding it up? i.e. need the MxNet team to make some change to the main library?

**Comment by vchuravy** [comment]
1. Yes that is the branch I have been working on
2. The branch has a small example in there and the best thing to do is to execute that and see it crash. Then look into where the issues are coming from.
3. There is one immediate issue that comes to mind and that is the livetime of objects in the mxnet side of things. Everything that is passed to Julia needs to be allocated on the heap and then free'd later on. Otherwise we will access data that is no longer valid. Python gets away with this because everything is synchronous. 

I will see if I can devote some more time to this, but work gets in the way right now.

**Comment by Andy-P** [comment]
@vchuravy ,

I forked your code and tried to run it but get a Segfaults when running the script julia-softmax.jl
Is this the same error you ran into?

```
.... 
mysoftmax = JuliaSoftmax()
info = mx.Native.NDArrayOpInfo(mysoftmax)
pstring = bytestring("0x", hex(reinterpret(UInt, pointer_from_objref(info))))

# this next line cause a segfault
mlp = mx._NDArray(name = "softmax", info = pstring, data=fc3)

```
_note: I'm not setup on this windows machine to capture segfaults, so a screen shot will have to suffice.. sorry :-|_

![segerror](https://cloud.githubusercontent.com/assets/831751/12877924/a64725d8-ce5e-11e5-92a4-9bdc5545faa6.jpg)

Version Info

```
Julia Version 0.4.2
Commit bb73f34 (2015-12-06 21:47 UTC)
Platform Info:
  System: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-4590T CPU @ 2.00GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

**Comment by Andy-P** [comment]
After rebuilding MXNet and  doing a Pkg.checkout("MXNet") I got a bit further. The error below is generated on calling the 'fit!()' function. Unfortunately my lack of C++ knowledge makes this difficult to diagnose.

I```
NFO: Initializing parameters...
INFO: Creating KVStore...
INFO: Start training...
INFO: == Epoch 001 ==========
INFO: ## Training summary
INFO:           accuracy = 0.1033

signal (11): Segmentation fault: 11
MXNDArrayFree at /Users/userme/.julia/v0.4/MXNet/deps/usr/lib/libmxnet.so (unknown line)
#4#delete! at /Users/userme/.julia/v0.4/MXNet/src/base.jl:53
jl_apply at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/./julia.h:1325
jl_apply at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/./julia.h:1410
run_finalizers at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/gc.c:362
_ZL11to_functionP17_jl_lambda_info_t at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/codegen.cpp:691
jl_compile at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/codegen.cpp:812
jl_trampoline_compile_function at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/builtins.c:1025
jl_apply at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/./julia.h:1325
jl_apply at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/gf.c:1691
anonymous at replutil.jl:23
with_output_limit at /Applications/Julia-0.4.2.app/Contents/Resources/julia/lib/julia/sys.dylib (unknown line)
writemime at replutil.jl:23
jlcall_writemime_22378 at  (unknown line)
jl_apply at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/gf.c:1691
showerror at replutil.jl:50
showerror at replutil.jl:83
julia_showerror_22374 at  (unknown line)
jlcall_showerror_22374 at  (unknown line)
jl_apply at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/gf.c:1691
anonymous at task.jl:165
with_output_color at util.jl:313
jl_apply at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/gf.c:1691
anonymous at task.jl:63
jl_apply at /Users/osx/buildbot/slave/package_osx10_9-x64/build/src/./julia.h:1325
Julia has stopped: null, SIGSEGV
```

**Comment by Andy-P** [comment]
@vchuravy @pluskid 

I spent some time this past week trying to debug this. Unfortunately, my lack of experience with c++ has  made progress very slow.

I have spent a fair amount of time testing MXNet.jl and it is easily the best solution for Julia + Deep Learning if you can get away with using the current native loss functions.  Unfortunately for me, this is not the case. So this issue of a custom layer has become THE issue for me. It is basically the only thing standing in the way of my using MXNet.jl for serious work. 

So I need to ask, is this something that will likely take a long time to solve either because it is actually quite difficult, or because no one who has the skills has the time?  Or is it just one of those things that just needs a little extra love? I really can't tell.

If it seems unlikely to be solve soon that is fine. I will look for another solution and come back to MXNet later. On the other hand, if it is just around the corner I will hold off investing time in a different framework and instead spend my time contributing where I can.

**Comment by vchuravy** [comment]
I hope that I can make another push for this during the coming week. There
are a few issues I have a handle on and one that I am not sure about. It
requires another serious push and so for me it is a time thing. But I am in
a similar position to you.

Keep in mind that this only gets us a CPU implementation and I haven't
looked at how the Python interface handles GPU (I know that it does), so
that might be something worthwhile to check out.

**Comment by CarloLucibello** [comment]
+1 for this, I'm really looking forward for it
